hazelcast/src/main/java/com/hazelcast/jet/impl/MasterJobContext.java

```diff
@@ -194,57 +200,89 @@ private boolean isCancelled() {
+                    DAG dag = resolveDag(executionIdSupplier);
+                    if (dag == null) {
+                        return null;
+                    }
+                    // must call this before rewriteDagWithSnapshotRestore()
+                    String dotRepresentation = dag.toDotString(defaultParallelism, defaultQueueSize);
+                    long snapshotId = jobExecRec.snapshotId();
+                    String snapshotName = mc.jobConfig().getInitialSnapshotName();
+                    String mapName =
+                            snapshotId >= 0 ? jobExecRec.successfulSnapshotDataMapName(mc.jobId())
+                                    : snapshotName != null ? EXPORTED_SNAPSHOTS_PREFIX + snapshotName
+                                    : null;
+                    if (mapName != null) {
+                        rewriteDagWithSnapshotRestore(dag, snapshotId, mapName, snapshotName);
+                    } else {
+                        logger.info("Didn't find any snapshot to restore for " + mc.jobIdString());
+                    }
+                    logger.info("Start executing " + mc.jobIdString()
+                            + ", execution graph in DOT format:\n" + dotRepresentation
+                            + "\nHINT: You can use graphviz or http://viz-js.com to visualize the printed graph.");
+                    logger.fine("Building execution plan for " + mc.jobIdString());
+                    return dag;
+                })
+                .thenComposeAsync(dag -> {
+                    if (dag == null) {
+                        return completedFuture(null);
+                    }
+                    MembersView membersView = Util.getMembersView(mc.nodeEngine());
+                    JobExecutionRecord jobExecRec = mc.jobExecutionRecord();
+
+                    ClassLoader classLoader = mc.getJetServiceBackend().getJobClassLoaderService()
+                                                .getClassLoader(mc.jobId());
+                    return Util.doWithClassLoader(classLoader, () ->
+                                       createExecutionPlans(mc.nodeEngine(), membersView.getMembers(),
+                                               dag, mc.jobId(), mc.executionId(), mc.jobConfig(), jobExecRec.ongoingSnapshotId(),
+                                               false, mc.jobRecord().getSubject()))
+                               .thenApply(executionPlan -> new StartJobInitExecutionParams(executionPlan, membersView));
+                }, coordinationExecutor)
+                .thenCompose(tuple -> coordinationService.submitToCoordinatorThread(() -> {
+                    if (tuple != null) {
+                        initExecution(tuple.membersView, tuple.plans);
+                    }
+                }))
+                .whenComplete((r, e) -> {
+                    if (e != null) {
+                        finalizeJob(peel(e));
+                    }
+                })
+                .join();
```

Comment:
Formerly, we didn't block the execution here, I believe we must not block here. It seems to be the cause of the deadlock in the `ParallelStressTest`.